File: src/main/scala/g0001_0100/s0076_minimum_window_substring/Solution.scala
package g0001_0100.s0076_minimum_window_substring

// #Hard #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table #Sliding_Window
// #Level_2_Day_14_Sliding_Window/Two_Pointer #Big_O_Time_O(s.length())_Space_O(1)
// #2023_11_02_Time_453_ms_(100.00%)_Space_54.7_MB_(100.00%)

object Solution {
    def minWindow(s: String, t: String): String = {
        val map = new Array[Int](128)
        t.foreach(char => map(char - 'A') += 1)

        var count = t.length
        var begin = 0
        var end = 0
        var d = Int.MaxValue
        var head = 0

        while (end < s.length) {
            if (map(s.charAt(end) - 'A') > 0) {
                count -= 1
            }
            map(s.charAt(end) - 'A') -= 1
            end += 1

            while (count == 0) {
                if (end - begin < d) {
                    d = end - begin
                    head = begin
                }
                if (map(s.charAt(begin) - 'A') == 0) {
                    count += 1
                }
                map(s.charAt(begin) - 'A') += 1
                begin += 1
            }
        }

        if (d == Int.MaxValue) "" else s.substring(head, head + d)
    }
}

File: src/main/dart/g0001_0100/s0076_minimum_window_substring/Solution.dart
// #Hard #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table #Sliding_Window
// #Level_2_Day_14_Sliding_Window/Two_Pointer #Top_Interview_150_Sliding_Window
// #Big_O_Time_O(s.length())_Space_O(1) #2024_10_05_Time_337_ms_(92.86%)_Space_150.5_MB_(57.14%)

class Solution {
  String minWindow(String s, String t) {
    List<int> map = List.filled(128, 0);

    for (int i = 0; i < t.length; i++) {
      map[t.codeUnitAt(i)]++;
    }

    int count = t.length;
    int begin = 0;
    int end = 0;
    int d = double.maxFinite
        .toInt();
    int head = 0;

    while (end < s.length) {
      if (map[s.codeUnitAt(end++)]-- > 0) {
        count--;
      }
      while (count == 0) {
        if (end - begin < d) {
          d = end - begin;
          head = begin;
        }
        if (map[s.codeUnitAt(begin++)]++ == 0) {
          count++;
        }
      }
    }

    return d == double.maxFinite.toInt() ? "" : s.substring(head, head + d);
  }
}
